### PR TITLE
Fix compilation with FP16_QK_REDUCTION enabled.

### LINKED ISSUE #806: [BUG] attention/prefill.cuh(138): error: expression must have a constant value
Hi, when I compile flashInfer C++ API from scratch. I met the following compile error. Can anyone help to have a look?
Thanks!

**Environment**:
CUDA version: cuda12.6
Host compiler: g++ (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
OS: Ubuntu 24.04
GPU: NVIDIA GeForce RTX 4050 Laptop

**Error Message**:
```
/data/work/flashinfer/include/flashinfer/attention/prefill.cuh(138): error: expression must have a constant value
        AttentionVariant::use_softmax ? DTypeQKAccum(-math::inf) : DTypeQKAccum(0.f);
        ^
/data/work/flashinfer/include/flashinfer/attention/prefill.cuh(138): note #2703-D: cannot call non-constexpr function "__half::__half(float)" (declared at line 4608 of /data/cuda-12.6/bin/../targets/x86_64-linux/include/cuda_fp16.h)
        AttentionVariant::use_softmax ? DTypeQKAccum(-math::inf) : DTypeQKAccum(0.f);
                                        ^
          detected during:
            instantiation of class "flashinfer::KernelTraits<MASK_MODE_, CTA_TILE_Q_, NUM_MMA_Q_, NUM_MMA_KV_, NUM_MMA_D_QK_, NUM_MMA_D_VO_, NUM_WARPS_Q_, NUM_WARPS_KV_, POS_ENCODING_MODE_, DTypeQ_, DTypeKV_, DTypeO_, DTypeQKAccum_, IdType_, AttentionVariant_> [with MASK_MODE_=flashinfer::MaskMode::kNone, CTA_TILE_Q_=16U, NUM_MMA_Q_=1U, NUM_MMA_KV_=1U, NUM_MMA_D_QK_=8U, NUM_MMA_D_VO_=8U, NUM_WARPS_Q_=1U, NUM_WARPS_KV_=4U, POS_ENCODING_MODE_=flashinfer::PosEncodingMode::kNone, DTypeQ_=half, DTypeKV_=__nv_fp8_e5m2, DTypeO_=half, DTypeQKAccum_=half, IdType_=int32_t, AttentionVariant_=flashinfer::AttentionVariant4]" at line 735

```
**Reproduce**:
```
/data/cuda-12.6/bin/nvcc -forward-unknown-to-host-compiler -DFLASHINFER_ENABLE_BF16 -DFLASHINFER_ENABLE_FP8_E4M3 -DFLASHINFER_ENABLE_FP8_E5M2 -I/data/work/flashinfer/include -O3 -DNDEBUG -arch=native -Xcompiler=-fPIC --fatbin-options -compress-all -MD -MT
CMakeFiles/prefill_kernels.dir/src/generated/batch_paged_prefill_head_qk_128_head_vo_128_posenc_0_fp16qkred_1_mask_0_dtypeq_f16_dtypekv_e5m2_dtypeout_f16_idtype_i32.cu.o 
-MF CMakeFiles/prefill_kernels.dir/src/generated/batch_paged_prefill_head_qk_128_head_vo_128_posenc_0_fp16qkred_1_mask_0_dtypeq_f16_dtypekv_e5m2_dtypeout_f16_idtype_i32.cu.o.d -x cu 
-c /data/work/flashinfer/src/generated/batch_paged_prefill_head_qk_128_head_vo_128_posenc_0_fp16qkred_1_mask_0_dtypeq_f16_dtypekv_e5m2_dtypeout_f16_idtype_i32.cu 
-o CMakeFiles/prefill_kernels.dir/src/generated/batch_paged_prefill_head_qk_128_head_vo_128_posenc_0_fp16qkred_1_mask_0_dtypeq_f16_dtypekv_e5m2_dtypeout_f16_idtype_i32.cu.o --std=c++17
```

### LINKED ISSUE #936: Kernels generated with use_fp16_qk_reductions=true break the LogitsTransform implementation used by prefill kernels
I want to clarify the semantics of the `LogitsTransform` function declared on :https://github.com/flashinfer-ai/flashinfer/blob/738460ff82e2230ebcc8dff50e49e1d6278e011a/include/flashinfer/attention/variants.cuh#L69

The `logits` parameter is templated and presumably can support `__half`. However, the computation of the value of `logits` on 

https://github.com/flashinfer-ai/flashinfer/blob/738460ff82e2230ebcc8dff50e49e1d6278e011a/include/flashinfer/attention/variants.cuh#L75 causes a compilation failure as the `operator *` cannot be resolved when the types are `__half` and `float`. Also, the assignment will also likely break because of the implicit `float` to `__half` conversion.

The `LogitsTransform` template is invoked inside the `prefill` kernel on line https://github.com/flashinfer-ai/flashinfer/blob/738460ff82e2230ebcc8dff50e49e1d6278e011a/include/flashinfer/attention/prefill.cuh#L690

I discovered this issue when working on fixing #806 and compiling kernels that were generated with the `use_fp16_qk_reductions=true` flag passed to `aot_build_utils.generate`.

I can apply a fix using a `constexpr` cast from fp16 to fp32 and _vice-versa_ either at call-site or inside `LogitsTransform` . But, before I do any mechanical changes wanted to clarify the intent of the implementation.

### PR DESCRIPTION
**Pull request by diptorupd**
As described in #806 and #936, setting the cmake build flag `FLASHINFER_GEN_USE_FP16_QK_REDUCTIONS` to "true" causes a build failure due to `cuda_fp16.h` not supporting `constexpr` cast from `__half` type to `float`. Note that the issue is not just a CMake/C++ configuration issue the issue will be triggered even in the flashinfer JIT code compilation path as reported in #915.

The PR fixes #806 and #936 by adding a modified version of the FP16 header from the [FP16 library](https://github.com/Maratyszcza/FP16) that supports `constexpr` versions of the conversion functions. To make the conversion functions `constexpr`, I am using `std::bit_cast` that is the reason for bumping the required standard to 20.

With these changes I am able to build the C++ API with both `FLASHINFER_GEN_USE_FP16_QK_REDUCTIONS` ON and OFF.

Fixes #936 
Fixes #806

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -157,21 +209,33 @@ file(GLOB_RECURSE DISPATCH_INC_FILE
...
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CXX_FLAGS}")
+set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} ${NVCC_FLAGS}")
+
 add_library(decode_kernels STATIC ${DECODE_KERNELS_SRCS})
 target_include_directories(decode_kernels PRIVATE ${FLASHINFER_INCLUDE_DIR})
-target_compile_options(decode_kernels PRIVATE -Xcompiler=-fPIC --fatbin-options
-                                              -compress-all)
+if(FLASHINFER_GEN_USE_FP16_QK_REDUCTIONS)
+  target_link_libraries(decode_kernels PRIVATE Boost::math)
```
Comment (yzh119) [inline CMakeLists.txt:230]: This is the dependency of `fp16.h` right?

**Review by yzh119 — APPROVED** [review]
LGTM, thanks for the contribution.

File: CMakeLists.txt (hunk 1 above)
Comment (diptorupd) [reply CMakeLists.txt:230]: Yes, that is right.
